File: C++/multiply-strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        vector<int> n1;
        vector<int> n2;

        if ((num1 == "0") || (num2 == "0")) {
            return "0";
        }
        
        for (char c: num1) {
            n1.push_back(c-'0');
        }
        for (char c: num2) {
            n2.push_back(c-'0');
        }

        int sz1 = n1.size();
        int sz2 = n2.size();
        vector<int> prod(sz1+sz2);
        for (int i = sz1 - 1; i >= 0; i--) {            
            for (int j = sz2 - 1; j >= 0; j--) {
                int p = n1[i] * n2[j] ;
                prod[i+j+1] += p;
            }
        }
        
        int carry = 0;
        for (int i = prod.size() - 1; i >= 0; i--) {
            int sum = prod[i] + carry;
            prod[i] = sum % 10;
            carry = sum / 10;
        }
        
        string s;
        int i = 0;
        while (prod[i] == 0)
            i++;
        for (; i < prod.size(); i++) {
            s += prod[i] + '0';
        }
        return s;
    }
};

File: Kotlin/multiply-strings.kt
class Solution {
    fun multiply(num1: String, num2: String): String {
        val n1 = num1.length
        val n2 = num2.length
        var tmp = IntArray(n1 + n2)
        for (i in 0 until n1) {
            for (j in 0 until n2) {
                tmp[i+j+1] += (num1[i] - '0') * (num2[j] - '0')
            }
        }
        var carry = 0
        for (i in tmp.size - 1 downTo 0) {
            tmp[i] += carry
            carry = tmp[i] / 10
            tmp[i] %= 10
        }
        var i = 0
        while (i < tmp.size && tmp[i] == 0) {
            i++
        }
        var ans = ""
        while (i < tmp.size) {
            ans += tmp[i].toString()
            i++
        }
        return if (ans == "") "0" else ans
    }
}